katran/lib/bpf/pckt_encap.h

```diff
@@ -221,4 +231,133 @@ decap_v4(struct xdp_md* xdp, void** data, void** data_end) {
+                                struct packet_description *pckt,
+                                struct real_definition *dst, __u32 pkt_bytes) {
+  void *data;
+  void *data_end;
+  struct iphdr *iph;
+  struct udphdr *udh;
+  struct eth_hdr *new_eth;
+  struct eth_hdr *old_eth;
+  struct real_definition *src;
+
+  __u16 sport = bpf_htons(pckt->flow.port16[0]);
+  __u32 ipv4_src  = V4_SRC;
+
+  src = bpf_map_lookup_elem(&pckt_srcs, &ipv4_src);
+  if (!src) {
+    return false;
+  }
+  ipv4_src = src->dst;
+ 
+  sport ^= ((pckt->flow.src >> 16) & 0xFFFF);
+  __u64 csum = 0;
+
+  if (bpf_xdp_adjust_head(
+      xdp, 0 - ((int)sizeof(struct iphdr) + (int)sizeof(struct udphdr)))) {
+    return false;
+  }
+  data = (void *)(long)xdp->data;
+  data_end = (void *)(long)xdp->data_end;
+  new_eth = data;
+  iph = data + sizeof(struct eth_hdr);
+  udh = (void *)iph + sizeof(struct iphdr);
+  old_eth = data + sizeof(struct iphdr);
+  if (new_eth + 1 > data_end ||
+      old_eth + 1 > data_end ||
+      iph + 1 > data_end ||
+      udh + 1 > data_end) {
+    return false;
+  }
+  memcpy(new_eth->eth_dest, cval->mac, 6);
+  memcpy(new_eth->eth_source, old_eth->eth_dest, 6);
+  new_eth->eth_proto = BE_ETH_P_IP;
+  
+  create_udp_hdr(
+    udh,
+    sport,
+    GUE_DPORT,
+    pkt_bytes + sizeof(struct udphdr),
+    GUE_CSUM);
+
+  create_v4_hdr(
+    iph,
+    pckt,
+    ipv4_src,
+    dst->dst,
+    pkt_bytes + sizeof(struct udphdr),
+    IPPROTO_UDP);
+
+  return true;
+}
+
+__attribute__((__always_inline__))
+static inline bool gue_encap_v6(struct xdp_md *xdp, struct ctl_value *cval,
+                                bool is_ipv6, struct packet_description *pckt,
+                                struct real_definition *dst, __u32 pkt_bytes) {
+  void *data;
+  void *data_end;
+  struct ipv6hdr *ip6h;
+  struct eth_hdr *new_eth;
+  struct eth_hdr *old_eth;
+  struct udphdr *udh;
+  __u32 key = V6_SRC;
+  __u16 payload_len;
+  __u16 sport;
+  struct real_definition *src;
+
+  src = bpf_map_lookup_elem(&pckt_srcs, &key);
+  if (!src) {
+    return false;
+  }
+ 
+  if (bpf_xdp_adjust_head(
+    xdp, 0 - ((int)sizeof(struct ipv6hdr) + (int)sizeof(struct udphdr)))) {
+    return false;
+  }
+  data = (void *)(long)xdp->data;
+  data_end = (void *)(long)xdp->data_end;
+  new_eth = data;
+  ip6h = data + sizeof(struct eth_hdr);
+  udh = (void *)ip6h + sizeof(struct ipv6hdr);
+  old_eth = data + sizeof(struct ipv6hdr);
+  if (new_eth + 1 > data_end ||
+      old_eth + 1 > data_end ||
+      ip6h + 1 > data_end ||
+      udh + 1 > data_end) {
+    return false;
+  }
+  memcpy(new_eth->eth_dest, cval->mac, 6);
+  memcpy(new_eth->eth_source, old_eth->eth_dest, 6);
+  new_eth->eth_proto = BE_ETH_P_IPV6;
+
+
+  if (is_ipv6) {
+    sport = (pckt->flow.srcv6[3] & 0xFFFF) ^ pckt->flow.port16[0];
+    pkt_bytes += (sizeof(struct ipv6hdr) + sizeof(struct udphdr));
+  } else {
+    sport = ((pckt->flow.src >> 16) & 0xFFFF) ^ pckt->flow.port16[0];
```

Comment:
In terms of entropy, wouldn't the first 16 bits be better than last 16 bits?

Reply:
no. last 16bits are usually clients mac address. first 16 would be the same for huge amount of networks (e.g. 2001::/16)